process/sync/basicForkDetector.go

```diff
@@ -81,16 +81,23 @@ func (bfd *basicForkDetector) AddHeader(
 
 	if state == process.BHProcessed {
 		// create a check point and remove all the past headers
-		if finalHeader == nil || finalHeader.IsInterfaceNil() {
+		if finalHeaders == nil {
 			bfd.setFinalCheckpoint(bfd.lastCheckpoint())
-		} else if finalHeader.GetNonce() > bfd.GetHighestFinalBlockNonce() {
-			bfd.setFinalCheckpoint(&checkpointInfo{nonce: finalHeader.GetNonce(), round: finalHeader.GetRound()})
-			bfd.append(&headerInfo{
-				nonce: finalHeader.GetNonce(),
-				round: finalHeader.GetRound(),
-				hash:  finalHeaderHash,
-				state: process.BHNotarized,
-			})
+		} else {
+			for i := 0; i < len(finalHeaders); i++ {
+				if finalHeaders[i].GetNonce() > bfd.GetHighestFinalBlockNonce() {
```

Comment:
Should extract finalHeaders[i].GetNonce() > bfd.GetHighestFinalBlockNonce() into a var

Reply:
Done. Also extracted all the code from for in new method